File: examples/tutorial_4_mimo.py
"""MIMO tutorial (v3.0) - Python port of ``Examples/Tutorial_4_MIMO.m`` (core).

Identifies a 2x2 modal plant from an orthogonal 2-input multisine:
  * non-parametric MIMO FRF with ``time2frf_ml`` (and ``time2frf_lpm``),
  * structured modal parameters with ``frf2modal``.
Self-contained synthetic 2x2 plant (the benchmark model is SISO).
"""

import numpy as np
import control
import matplotlib.pyplot as plt

import fditools as fdi
from _plot import save_fig, show


def main():
    # --- 2x2 modal plant (two resonances, with cross-coupling) ----------
    w1, w2 = 2 * np.pi * 60, 2 * np.pi * 150
    g = [[control.tf([w1 ** 2], [1, 2 * 0.02 * w1, w1 ** 2]),
          control.tf([0.3 * w1 ** 2], [1, 2 * 0.05 * w1, w1 ** 2])],
         [control.tf([0.2 * w2 ** 2], [1, 2 * 0.04 * w2, w2 ** 2]),
          control.tf([w2 ** 2], [1, 2 * 0.03 * w2, w2 ** 2])]]
    nu = ny = 2

    # --- orthogonal 2-input multisine -> 2 experiments ------------------
    harm = dict(fs=5000.0, df=1.0, fl=5.0, fh=400.0, fr=1.02)
    ms = fdi.multisine(harm, [control.tf([1], [1]), control.tf([1], [1])],
                       dict(itp="r", ctp="n", dtp="f", gtp="l"))
    ne = ms.x.shape[1]
    nrofp = 6
    N = nrofp * ms.nrofs
    T = np.arange(N) / harm["fs"]
    rng = np.random.default_rng(0)

    Xl, Yl = [], []          # transient-removed (for ML)
    Xraw, Yraw = [], []      # raw (for LPM)
    for e in range(ne):
        Ue = np.tile(ms.x[:, e, :].T, (nrofp, 1))
        Ye = np.zeros((N, ny))
        for o in range(ny):
            for i in range(nu):
                Ye[:, o] += control.forced_response(g[o][i], T, Ue[:, i]).outputs
            Ye[:, o] += 1e-4 * rng.standard_normal(N)
        Xraw.append(Ue)
        Yraw.append(Ye)
        Xl.append(fdi.pretreat(Ue, ms.nrofs, harm["fs"], 1, 0)[0])
        Yl.append(fdi.pretreat(Ye, ms.nrofs, harm["fs"], 1, 0)[0])

    # --- non-parametric MIMO FRF ----------------------------------------
    Pml = fdi.time2frf_ml(np.stack(Xl, 2), np.stack(Yl, 2), ms)
    Plpm = fdi.time2frf_lpm(np.stack(Xraw, 2), np.stack(Yraw, 2), harm["fs"],
                            order=2, halfwidth=3, period=ms.nrofs, lines=ms.ex)
    print(f"MIMO FRF: {Pml.response.shape}  method={Pml.userdata.method}")

    freq = Pml.freq
    true = np.array([[[complex(g[o][i](1j * 2 * np.pi * f)) for f in freq]
                      for i in range(nu)] for o in range(ny)])
    print("ML  median rel err:", f"{np.median(np.abs(Pml.response - true) / np.abs(true)):.2e}")
    print("LPM median rel err:", f"{np.median(np.abs(Plpm.response - true) / np.abs(true)):.2e}")

    # --- structured modal identification --------------------------------
    modal, Pm = fdi.frf2modal(Pml, 0, 2, damping="proportional",
                              initfreq=[60, 150], initdamp=0.02, feedthrough=False)
    print("modal wn [Hz]:", np.round(np.sort(modal["wn"]), 2), " (true [60, 150])")
    print("modal zeta   :", np.round(modal["zeta"][np.argsort(modal["wn"])], 4),
          " (true [0.02, 0.03])")

    # --- plot the (1,1) channel -----------------------------------------
    modal_resp = np.array([complex(control.tf(control.ss(Pm)[0, 0])(1j * 2 * np.pi * f))
                           for f in freq])
    fig, _ = fdi.bode_fdi(
        [(freq, true[0, 0]), Pml[0, 0], (freq, modal_resp)],
        legend=["true", "FRF (ML)", "modal fit"],
        title="MIMO 2x2 identification - channel (1,1)")
    save_fig(fig, "tutorial_4_mimo.png")
    show()


if __name__ == "__main__":
    main()

File: MATLAB/Examples/Tutorial_4_MIMO.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% TUTORIAL OF MIMO FRF MEASUREMENT
% --------------------------------
% Descr.:   2-input / 2-output FRF identification on the shared MIMO benchmark
%           mimobench (the SAME 2x2 rank-one modal plant used by the Step_MIMO
%           series, modes ~40 / 95 / 180 Hz). Two excitation strategies are
%           demonstrated and compared against the true plant:
%             (A) orthogonal multisine over MULTIPLE experiments
%             (B) ZIPPERED multisine in a SINGLE experiment
%           The orthogonal FRF then feeds structured modal identification
%           (frf2modal), proportional and general-viscous damping.
% System:   2x2 rank-one modal benchmark (mimobench), shared with Step_MIMO*.
% Author:   Wataru Ohnishi, The University of Tokyo, 2026
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear all; close all; clc;
[G0, wn_true, ze_true] = mimobench();        % true 2x2 plant (shared benchmark)

% Common experiment settings (aligned with the Step_MIMO series)
harm.fs = 2500;     harm.df = 1;             % sampling / resolution
harm.fl = 1;        harm.fh = 500;           % excited band (1 Hz - 500 Hz)
harm.fr = 1.02;                               % (qlog ratio, unused for 'l')
nrofp = 5;          trans = 1;                % periods kept / transient periods
nptot = nrofp + trans;
outnoise = 1e-3;                              % output noise (sets the FRF floor)

%% ===== METHOD A: ORTHOGONAL MULTISINE, MULTIPLE EXPERIMENTS =====
% multisine(nrofi=2) builds 2 orthogonal (Hadamard) experiments: in each
% experiment BOTH inputs are driven with phase patterns that make the input
% matrix U(2x2) invertible at every frequency, so G = Y/U gives the full 2x2.
options.itp = 'r'; options.ctp = 'c'; options.dtp = 'f'; options.gtp = 'l';
nrofi = 2;
Hampl = repmat(tf(1),[1,nrofi]);
ms    = multisine(harm, Hampl, options);
nrofs = ms.nrofs;
nexp  = nrofi;                              % orthogonal -> nrofi experiments

rng('default');
t  = (0:nptot*nrofs-1)'/harm.fs;
uc = cell(1,nexp); yc = cell(1,nexp);
for j = 1:nexp
    Uj  = squeeze(ms.x(:,j,:)).';          % one period, nrofs x 2 (experiment j)
    in  = repmat(Uj, [nptot,1]);
    out = lsim(G0, in, t);
    out = out + outnoise*randn(size(out));
    uc{j} = in;  yc{j} = out;
end
% multi-experiment iodata (cell per experiment); estimator auto-detects MIMO
datA = iodata(yc, uc, 1/harm.fs, 'Period', nrofs, 'UserData', struct('ms',ms));
datA = pretreat(datA, 'trans', trans);
Pa   = time2frf_ml(datA);                  % full 2x2 frd

figure('Name','MIMO FRF: orthogonal multiple experiments');
G0a = frd(freqresp(G0, 2*pi*Pa.Frequency), Pa.Frequency, 'FrequencyUnit','Hz');
bode(G0a, Pa); grid on;
legend('true','estimated (orthogonal, 2 exp)','Location','best');
title('2\times2 FRF - orthogonal multisine, multiple experiments');

%% ===== METHOD B: ZIPPERED MULTISINE, SINGLE EXPERIMENT =====
% One experiment, but each input excites DISJOINT (interleaved) lines:
% input 1 -> odd excited lines, input 2 -> even ones. At any excited line only
% one input is active, so each FRF column is read off directly. Full 2x2 from a
% single record, at the cost of halved per-channel frequency resolution.
nrofsZ = round(harm.fs/harm.df);
fbin   = (0:nrofsZ/2-1)*harm.df;           % freq of bin k (1-based) = (k-1)*df
inband = find(fbin >= harm.fl & fbin <= harm.fh);   % 1-based bin indices
ex1 = inband(1:2:end);  ex2 = inband(2:2:end);      % zipper the lines

U  = zeros(nrofsZ,2);
ph1 = exp(1i*2*pi*rand(numel(ex1),1));
ph2 = exp(1i*2*pi*rand(numel(ex2),1));
U(ex1,1) = ph1;  U(nrofsZ-ex1+2,1) = conj(ph1);     % conjugate-symmetric ->
U(ex2,2) = ph2;  U(nrofsZ-ex2+2,2) = conj(ph2);     % real time signals
uz = real(ifft(U));
uz = uz ./ max(abs(uz),[],1);              % normalize each input amplitude

inz  = repmat(uz, [nptot,1]);
tz   = (0:size(inz,1)-1)'/harm.fs;
outz = lsim(G0, inz, tz);
outz = outz + outnoise*randn(size(outz));

ms_zip = struct('harm',harm, 'nrofs',nrofsZ, ...
                'ex',sort([ex1(:);ex2(:)]), 'freq',fbin(:));
datB = iodata(outz, inz, 1/harm.fs, 'Period', nrofsZ, 'UserData', struct('ms',ms_zip));
datB = pretreat(datB, 'trans', trans);
Pb   = time2frf_ml(datB);                  % 2x2 frd (each column on its lines)

figure('Name','MIMO FRF: zippered single experiment');
G0b = frd(freqresp(G0, 2*pi*Pb.Frequency), Pb.Frequency, 'FrequencyUnit','Hz');
bode(G0b, Pb); grid on;
legend('true','estimated (zippered, 1 exp)','Location','best');
title('2\times2 FRF - zippered multisine, single experiment');

% NOTE: the orthogonal method gives the full 2x2 at every line. The zippered
% method uses a single record but each column is measured only on its own
% (every-other) lines; time2frf_ml interpolates each column onto the full grid
% so it plots as a continuous curve - the true per-channel resolution is half.

%% ===== STRUCTURED MODAL IDENTIFICATION (van der Hulst et al., MSSP 2026) =====
% mimobench IS a proportionally-damped, rank-one modal system, so frf2modal can
% recover the modal parameters directly from the orthogonal FRF estimate Pa
% (exactly as in Step_MIMO5).
[modal, Pm] = frf2modal(Pa, 0, numel(wn_true), 'initfreq', wn_true*1.05);

fprintf('\n--- identified modal parameters ---\n');
fprintf(' mode |  wn_true   wn_est [Hz] |  z_true    z_est\n');
for i = 1:numel(wn_true)
    fprintf('  %2d  | %8.2f  %8.2f    | %6.3f   %6.3f\n', ...
        i, wn_true(i), modal.wn(i), ze_true(i), modal.zeta(i));
end

% compare on the FRF grid so the axes match the data
G0m = frd(freqresp(G0, 2*pi*Pa.Frequency), Pa.Frequency, 'FrequencyUnit','Hz');
Pmf = frd(freqresp(Pm, 2*pi*Pa.Frequency), Pa.Frequency, 'FrequencyUnit','Hz');
figure('Name','Structured modal identification');
bode(G0m, Pa, Pmf); grid on;
legend('true','nonparametric FRF','identified modal model','Location','best');
title('2\times2 structured modal identification (frf2modal)');

% complex-FRF fit of the identified modal model vs the true plant
g0 = G0m.ResponseData(:);  gm = Pmf.ResponseData(:);
fprintf('modal model FRF fit vs true : %.2f %%\n', 100*(1-norm(gm-g0)/norm(g0-mean(g0))));

% --- general-viscous damping option (complex mode shapes, eqs.(2),(6),(46)) ---
% For non-proportionally damped systems use 'damping','general'; it also fits
% proportionally damped data (the complex mode shapes then come out ~real).
[modalG, PmG] = frf2modal(Pa, 0, numel(wn_true), ...
                          'damping','general', 'initfreq', wn_true*1.05);
PmGf = frd(freqresp(PmG, 2*pi*Pa.Frequency), Pa.Frequency, 'FrequencyUnit','Hz');
gG = PmGf.ResponseData(:);
fprintf('general-damping wn_est [Hz] :'); fprintf(' %.2f', modalG.wn); fprintf('\n');
fprintf('general modal FRF fit vs true: %.2f %%\n', 100*(1-norm(gG-g0)/norm(g0-mean(g0))));
